Provide visibility for an active fork

ickc opened the pull request:
While this is unlikely to be merged, this PR provides some visibility to an active fork that has merged many other PR from here.

smarnach commented:
I'm not really sure what to do about this library. I currently don't have time to maintain it, but I'm happy to spend some time for a proper handover if we can find a new owner for the code. (I spun this off a StackOverflow answe and never meant to really maintain it, but it unexpectedly became somewhat popular.)

@sylikc Would you like to become the "official" owner of the code? Would anyone else like?

Things we should do to hand this over:

* Add a link to the maintained fork in the README of this repo.
* Try to get ownership of the PyPI packages someone created for this.
* Get external links updated, e.g. the one from my StackOverflow answer.
* Update the documentation link in the README to no longer point to the documentation hosted here.

ickc commented:
There’s also a way to transfer in GitHub to make his “official” and others forks of his. I forgot exactly how but it can be done.

ickc commented:
I made 3 PR to @sylikc ~11 days ago. But there's no activity there. Looking at @sylikc/pyexiftool I found that all the new commits happened in July 2019 so I'm not sure if he's still actively maintaining that.

I recently wrote a tool (so far for personal use only) that depends on pyexiftool. It is very likely that I'll continue to use my tool and hence continue to improve pyexiftool since there's seem no other better way (my tool needs to know the metadata from raw so exiftool is uniquely capable on that.) And unless I move to another language or write my own code to basically does what pyexiftool does then I should be able to continue to maintain that.

And I currently maintained a few packages on PyPI so it is likely I can continue to keep the packages up-to-date in PyPI including CI on newer versions of Python. (I might add a conda package for it too since for conda environment it is more convenient.)

I think if @sylikc wants to, since he has longer history on using that tool and already done a dozen merges on the PRs here, he should be the first choice. But if we don't hear from him soon we may interpret that he's not actively maintaining that either. Then I'm willing to pick it up.

sylikc committed:
Merge branch 'support-w' of https://github.com/ickc/pyexiftool into ickc-support-w

as per manual merge instructions

sylikc committed:
Merge branch 'ickc-support-w'

adding ickc feature for no_output from GitHub PR #3

sylikc commented:
I am happy to maintain this.  When I have free time, I'm actively developing my own scripts to accelerate my photography workflow.  I have an internal gitlab so haven't been as active checking GitHub thinking no one would use it :)

Thanks for the enhancements @ickc  and thanks for making the code available and doing all the initial legwork @smarnach 

I'm aware one of the forks in some way or form is in PyPI which of course is many versions behind what PR I've merged from this project.

sylikc commented:
I don't have the experience with PyPI, but I'd be willing to learn from you @ickc how to get the package up, and maintain it.

sylikc committed:
version bump because of the merged PRs

ickc commented:
Sounds like a plan. If @smarnach agree then may be @smarnach could transfer the ownership of GitHub to @sylikc first?

About PyPI, currently on PyPI PyExifTool is owned by intense.feel/Martin Carnogursky in <https://pypi.org/user/intense.feel/>. Do you know his GitHub username? If we can't contact him then we could just have it in an alternative name (not uncommon when they have multiple forks.)

RootLUG commented:
That's me on PyPI, I was going to upload the forked & modified version but forgot to change the package name before uploading so I took the package name by accident. Please let me know the PyPI name of the new maintainer and with a confirmation from @smarnach I will transfer the ownership of the package on PyPI.

sylikc commented:
I created an account on PyPI to prepare for a transfer.  same username.

smarnach commented:
@sylikc @RootLUG Thanks a lot, sounds all good to me. It will take me a few days until I can find time to update links etc, but I will eventually get there.

sylikc commented:
cool @smarnach how do I set up the githubio page for the documentation?  is there an automated process to generate the docs?

csparker247 committed:
Fix NoneType error in common args check

smarnach commented:
Sorry about the delay! This is still on my list, but I'm really busy at the moment. :-/

sylikc commented:
Is there any update on this @smarnach ?

@ickc I've reviewed other PyPI modules and there's no standard for renaming things to be py-something so I'm going to roll back that specific commit.  I'm going to be working with the library again since we're all stuck at home... I'll try not to break any interfaces if I add features ;)

sylikc committed:
Revert "0.4.0 rename to pyexiftool and make sure tests pass"

This reverts commit de22a9624c4ff256a05616fbded54494af86b61e.

This reverts the rename of "exiftool" to "pyexiftool" at the suggestion of the community.  The exiftool name will remain for all of eternity.

sylikc committed:
Merge branch 'fix-nonetype-exception' of https://github.com/csparker247/pyexiftool into csparker247-fix-nonetype-exception

resolve bug in using the wrong common_args in the constructor.  in Pull request #5 by @csparker247

sylikc committed:
Merge branch 'csparker247-fix-nonetype-exception'

sylikc committed:
start the initial legwork of breaking the single exiftool.py into logical chunks in a directory.

This should help with code review clearer moving forward.  I'm not a big fan of huge source files...

The static helper functions (which rarely/should not change) will be isolated from the ExifTool-specific class

(tests pass)

sylikc committed:
add a comment at this line which uses a Python 3.3+ exclusive feature... so when it crashes here, there's no guessing why

sylikc committed:
put an extra TODO as we were debugging the code... might also need to check some type of errorlevel from exiftool

JSpenced commented:
@smarnach Any news on this? Be nice to have the forked version in PyPI. Thanks for all the effort so far.

melyux commented:
@smarnach please

RootLUG commented:
I see that there are multiple repos being forked with some of them more updated than others in terms of pull requests.
So let's maybe form a plan or otherwise we can discuss for another year before moving anywhere forward with this issue.
my proposal:
let's first agree on what repo should the code continue forward and would be the authority to publish the package on pypi. Is it this one or some of the forks mentioned in the issue? Maybe the forks are also very different and we should create a new repo where we rebase and merge all the changes from those forks + pending pull requests etc; basically, prepare the code for release. 
I can then very easily add a CI pipeline that would auto-publish the package on pypi based on whatever is in master branch on that repo that we would agree on. At that point, we can formulate some contribution guidelines and maintenance of the code (incoming pull requests) so we can get regular updates which would be automatically released via that CI pipeline.

That's of course just my opinion so if you guys have better plan, objections or suggestions please share 😄

RootLUG commented:
regarding the documentation, to me it looks like normal github-pages with static HTML. You can see it in https://github.com/smarnach/pyexiftool/tree/gh-pages so if you just generate sphinx docs into gh-pages branch it would publish that on github.io as static page.

ickc commented:
When I evaluated it 10 months ago, the sylikc repo is most mature. He basically merged all the PR in this repo into his, and reflected in his CHANGELOG.md. There was a questionable renaming in his 0.4 but after communicating with him he rolled that back in 0.4.2.

I think his fork is pretty solid, unless if you have something better and can point us to look at it.

I think doing the ownership transfer and set up the necessary CI and PyPI is not the problem—they are very easy to do.

I think @smarnach already agreed to this. The hard work is how and when it happens.

1. [transfer GitHub ownership of the repo](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/transferring-a-repository)
2. PyPI ownership transfer

Frankly I don't quite care about (2). Multiple forks of a package on PyPI is not new (even pyyaml has this.) (1) is much more important to inform people which repo is now the "authoritative" one to use. (1) can be easily done. Regarding the PyPI version, to me it suffices to say "package A is deprecated and please upgrade yours by installing its drop-in replacement by `pip uninstall ... && pip install ...`."

I think @sylikc's intention is to have a drop-in replacement of this as can be shown by his effort to rolled the back change such that no changes on the user-side has to happen.

sylikc commented:
I'm happy to continue to improve PyExifTool .  @RootLUG what other forks have features that I could/should pull into my fork?  

With @ickc 's direction, I certainly want to maintain compatibility as much as possible and keep it a drop-in replacement for the library.  As I edit/add features/calls, I am certainly mindful of not changing the existing behavior.

So, I can change the description on my project to say "up to date" or something.  I guess I could set up another PyPI to have it installable through pip?  Is that the preferred direction?

ickc commented:
Just to clarify, I think unnecessary breaking change should be avoided, but in the (unlikely) event that a breaking change is needed, a clear communication with a major version increment would suffices.

My suggestion given the amount of time has passed already, would be to start a new PyPI repo maintained by @sylikc (with whatever different name deemed appropriate.) When @smarnach has the time then the only thing he need to do is to change the ownership of this repo to @sylikc. In the meanwhile, I hope this PR has given that fork a visibility enough for people to migrate smoothly.

sylikc commented:
Got it.  Yeah, I'll keep that in mind.

Anyways, a new PyPI repo, _any suggestions on names_?  Or I'll go with the super imaginative **PyExifTool2** :D

DimmKirr commented:
@sylikc hey, thanks for taking a lead on that. Maybe for the name you could just use `exiftool` ?

asielen committed:
Added support for config files as outlined here: https://exiftool.org/config.html

Also added an optional params argument to the get_metadata method. This can be used to support output formating flags

asielen committed:
Added support for config files as outlined here: https://exiftool.org/config.html

Also added an optional params argument to the get_metadata method. This can be used to support output formating flags

sylikc commented:
Ah ha!  I _JUST_ noticed that I and @RootLUG have are assigned to be Maintainers of the PyExifTool project on PyPI!  Thanks very much @smarnach .  Let me see if I can get a release out soon

Actually, it looks like I was added awhile back lol.  and the project is owned by @RootLUG .  Oh well, same difference.

Still waiting for a ownership transfer of the repo though: https://docs.github.com/en/github/administering-a-repository/transferring-a-repository

sylikc committed:
v0.4.4 - ready for release to PyPI

sylikc committed:
fix formatting error in README.rst ... I'm new to a non .md format!

sylikc committed:
v0.4.5 more setup changes to get PyPI package in order

sylikc commented:
Alright, I look forward to an ownership transfer of repository... one day

but in the meantime, I think I've figured out how to release the package.  0.4.5 is now the PyPI release... I fixed the exclusion of tests ... please test and make sure this package is in order.

I made a bunch of changes (my first published project) to the setup.py and such so it would dump out a description.  I tested briefly in my environment, and it seems ok, but please do test.

sylikc commented:
starting discussions for direction of PyExifTool

Python 3.x in https://github.com/sylikc/pyexiftool/discussions/9
Planned re-design: https://github.com/sylikc/pyexiftool/discussions/10

Please discuss future topics there, and ideas.  I'm also looking at other projects which may have wrapped exiftool as well and see what common functionality people usually use.  I personally do a lot of metadata manipulation that I wish were a bit easier...

sylikc committed:
v0.5.0-alpha.0 new branch for upcoming changes.  See CHANGELOG.md for more information

sylikc committed:
breaks compatibility of constructor.  executable_ -> executable

fix inconsistent naming of class variables based on python conventions.  use of _ (underscores) to designate private variables, and attributes to get and set attributes rather than direct with variable
moving some of the default set lines to the top to see the defined class variables available to us
added some comment breaks so the code is at least more readable to me
added a "constants" file so we can refer to constants outside of the exiftool.py file

sylikc committed:
get the tests to pass in the way that makes sense

move the setting of the executable to be a property set, which can be changed
restore the behavior which you can read-only the "running" property, but prevent setting it

added test of attribute "running" and "executable"

sylikc committed:
be explicit in the warning returned in start()
make platform a constant so that there won't be a chance of typos comparing it in code

sylikc committed:
make the rest of the class consistently named for now

sylikc committed:
add some TODOs to check code and consider whether to refactor later
fixed self._process deletion.  To be consistent, it is always defined, and will be None if the process is not running

sylikc committed:
make block_size a property which can be set to define a read block size

sylikc committed:
as per comment https://gist.github.com/techtonik/4368898#gistcomment-2344225 , remove the Python 2.x code with one liner (requires Python 3.3)

sylikc committed:
change the targeting of Python version to 3.6+

sylikc committed:
as per discussion https://github.com/sylikc/pyexiftool/discussions/10#discussioncomment-503125 , change .start() to .run() to match the way subprocess library does it. (standardize the naming)
fix all tests to pass
remove the test that tests the find_executable() as it has been replaced with the standard library shutil.which()

sylikc committed:
(temporary until decided) return None right now when JSON result comes back empty.
add warning when terminating when Exiftool isn't running
and now because of that, check before calling terminate()

sylikc committed:
make the execute() work with a synchronization number.

This ensures that the return of {ready} is absolutely correct... in case a tag reads out "{ready}" or something, this is just a pre-emptive improvement

sylikc committed:
added code to read from stderr as well.  I think I'll be able to detect errors better by having an STDERR stream

This has not yet been tested on LINUX.  I will write tests later which tests this functionality in full on linux

sylikc committed:
change how terminate() works on Windows when __del__ is in progress.  communicate() freezes and never finishes execution in this case on win32 platform

sylikc committed:
change the function signature of terminate() to match subprocess API, timeout instead of wait_timeout

sylikc committed:
change the way ExifTool.running is detected.  In the event that exiftool dies or is killed, this is now detected by the property, rather than just relying on self._running

sylikc committed:
fixed the bug that freezes on windows because stderr is empty on os.read() by telling exiftool to print out a synchronization token on stderr at the end of processing.  This is similar to the {ready} sequence.  This ensures there is ALWAYS data on stderr, and in turn, Windows will no longer block on the stderr read

moved out the code for the fd read to a static function so we can use the same one for stdout and stderr

sylikc committed:
extracted all high-level functionality out of the ExifTool class.  the ExifToolHelper class currently contains the previous functionality in a verbatim way.  No functionality was changed.

sylikc committed:
consolidate the get_metadata functionality to take either a str or a list(str) and ALWAYS return a list straight from execute_json.  This fixes a bug that if a wildcard is passed as the input to filename, the old way would just return the first file ([0]) which is incorrect.

sylikc committed:
fixed a bug where if the filename is a number (it doesn't happen often but it can), filename is an int, and the the fsencode blows up

sylikc committed:
remove get_tags and rename get_tags_batch to get_tags, similar to the get_metadata.
changed the way to check for an iterable, so that you can pass arbitrary iterables to the function

sylikc committed:
fix tests to isolate the ExifTool and ExifToolHelper tests
changes the way tests are run against both the base class and the helper class

sylikc committed:
moved for extraneous functionality over from exiftool to helper

sylikc committed:
fixed the tests and now it's entirely working on Windows with the kill()

sylikc committed:
merge the handling of of get_metadata and get_tags... since they essentially do the same thing.  get_metadata is a special case alias of get_tags

sylikc committed:
fix the warning on helper_test so it doesn't terminate if not running

sylikc committed:
Merge pull request #7 from asielen/master

Support for config files.  There is a bug that will be fixed in the next commit

sylikc committed:
fix the bug as discussed on Pull request #7

sylikc committed:
Merge branch 'master' into v0.5.x-py3-refactor

merge the conflicts since the code has changed quite a bit and
merge in the changes into helper.py instead of the main ExifTool class

# Conflicts:
#	exiftool/exiftool.py

sylikc committed:
fix small bugs in the merging and make sure all tests pass

sylikc committed:
add the test that warns when the process dies after it was detected as running

sylikc committed:
added the config_file property to do error checking
changed the config_file setting code, to use the property.setter to do error checking
added a check of exiftool running prior to clearing the .run() method

sylikc committed:
commented out the fscodec() code that was first introduced in commit 843895ec6 back in 2012

I was unable to find the reference to where the code comes from the cpython source tree.  I would anticipate the old code from that commit was to deal with Python 2's str and unicode types that are no longer relevant.

tests still all pass.  If it ever is a problem, can add the code right back in since it was just commented out, not removed

sylikc committed:
as per recommendations in consistent testing, I wrote some Windows batch runner scripts to test various aspects... someone can help write the equivalent Linux scripts (or I probably will at some point)

These should help CI/CD later

sylikc committed:
start with legwork of getting annotations and such working with mypy, and using monkeytype to get some initial recommendations on typing

sylikc committed:
mypy helped fix a bug in terminate() referring to a bad variable in TimeoutExpired clause

sylikc committed:
simplify the format of the sequences with Python 3.6 f-strings

sylikc committed:
more type hints

sylikc committed:
added stub for logging

no functional changes otherwise
rearranged the layout of the methods to be more organized, and makes more sense (easier to find methods/properties)
rearranged __init__ code to make it explicit the variable declarations and the ones set by property

pulled out the @staticmethod to the module-level since it's really not a method tied to ExifTool... only used by it.  it doesn't use the class/instance, so Pythonic way is to have it as a module function

trim trailing spaces across the board

passed the test suite, meaning none of the changes above broke anything

sylikc commented:
Alright, doesn't appear that any activity is happening over here.  I'm going to tag all the people who've replied here to ask if you could review some of the code in my Python 3.x branch.

Since you're all users of PyExifTool, I presume, and to the very least, all changes I make, I want your valuable community input.

Please review the `v0.5.x-py3-refactor` branch, and welcome to discuss it on https://github.com/sylikc/pyexiftool/discussions/10 

@RootLUG @JSpenced @melyux @AutomationD @asielen 

Thanks!

sylikc committed:
add in empty placeholder for the tests/tmp dir

sylikc committed:
update gitignore for the tests/tmp/ dir

sylikc committed:
convert helper to using pathlib.Path for better clarity.
also, remove extraneous imports from test_exiftool

sylikc committed:
update tests for more coverage

sylikc committed:
fix pytest-cov script to only show coverage for exiftool, and not other included libraries from virtualenv

sylikc committed:
exiftool logger is still a WIP, not really implemented yet
add some testing for more coverage
created a windows.coveragerc and some comments to reduce clutter on things that can never get coverage (Linux-related lines)

sylikc committed:
add support for lists in tags.items in set_tags_batch()
(check CHANGELOG.md for more information)

fixes GitHub issue #12 , code contributed by @davidorme in https://github.com/sylikc/pyexiftool/issues/12#issuecomment-821879234

sylikc committed:
Merge branch 'master' into v0.5.x-py3-refactor

merge the set_tags_batch() allows lists in tags.items() into helper.py

# Conflicts:
#	CHANGELOG.md
#	exiftool/exiftool.py
#	setup.py

sylikc committed:
pulled out the @staticmethod to the module-level since it's really not a method tied to ExifToolHelper... only used by it.  it doesn't use the class/instance, so Pythonic way is to have it as a module function

added a stub to eventually write a keywords list test, but I'm not ready to do testing on it, so do it later

removed extraneous spacing

sylikc committed:
fixed an indent bug (from the merge) and annotation bug in helper.py

sylikc committed:
exiftool.py:
* change the way common_args works.  Before it was always -G -n, when it was None to get the default -G -n... now it's just in the constructor
* common_args is now set as a property, allowing you to interface with it directly even after creating the object
* stub was created to get and verify exiftool version, so that it can be retrieved in one go after launching the process.  this will also help santiy check you're running the right thing, not just some random process called "exiftool"

sylikc committed:
rename the 'doc' to 'docs' to follow more with the convention of the other folder names

sylikc committed:
trying to get documentation working again, need to document as we move forward with 0.5.0 build, since the changes are drastically different than v0.4 and below

sylikc committed:
added the version and version_tuple properties which is the exiftool version... not the class version

sylikc committed:
adjust copyright
put a TODO regarding the version check here... I think there's a bug with the checking, but it's probably solved by the version_tuple behavior

sylikc committed:
add some rudimentary logging into ExifTool class
you have to pass in a pre-configured logger, and ExifTool class will just use the passed in logger

sylikc committed:
put in some stub for Logger testing, but I encountered a freeze when zombie processes are around.
apparently after executing the _parse_ver the exiftool process remains open and doesn't die with main process when main process is killed (on windows exiftool.exe unpacks and launches another exiftool.exe)
investigating possible solutions

Commit:
update url for documentation

*.github.com has been replaced with *.github.io

sylikc committed:
Merge pull request #16 from beng/master

update url for documentation, contribution by @beng (Benjamin G)

sylikc committed:
Merge branch 'master' into v0.5.x-py3-refactor

merge the config PR #16 from @beng for updated link.  Merge it now so that later when I change it I don't have to deal with conflicts

sylikc committed:
added some other extensions to sphinx to generate cooler documentation (hopefully) - testing

add in a fix from @csparker247 regarding resolving the path
reference https://github.com/sylikc/pyexiftool/pull/13#discussion_r625888195

sylikc committed:
created a little private method that just unsets all the variables pertaining to running()

this should help in the future if we decide to put more things in, like say stdout or stderr clearing

jangop committed:
Encode parameters

sylikc committed:
removing now-defunct devnull code, since we now read from stderr and direct exiftool to output at least something to stderr

(this code was first commented on in commit 56af0a94e2b6a5e90a4583166eb2e02ff57aa832 .  Check the diff with the commit before it to see what it was like originally)

sylikc committed:
remove the logging.debug()

... this is now handled directly by setting the logger.  No extraneous logging

sylikc committed:
unified the call to subprocess.Popen so it's clearer in someways that there's no difference between the two platforms except the additiona kwargs

remove the optional parameter from set_pdefathsig(), since it's probably better to be explicit here
made the constant ALL CAPS, and for consistency, pulled it into constants.py

beautified the constants.py with block comments

sylikc committed:
super duper hacky fix to some Python interpreter vs exiftool interaction on windows only.
because a command is ran, the .kill() doesn't terminate the whole tree... this is probably not the best fix

jangop committed:
Test copying of tags

sylikc committed:
removed unnecessary check, as get_tags() has this parameter checking built-in

jangop committed:
Test basic functionality

All four implemented tests currently fail.

sylikc committed:
Merge pull request #19 from jangop/master

* copy_tags() parameters bugfix
* TestTagCopying test case

contribution by @jangop (Jan Philip Göpfert)

sylikc committed:
v0.4.9 update changelog and setup.py to reflect the previous two PR that got merged in, even if 0.4.8 was minor and didn't have any functional changes
further changed the links in the README.rst as sylikc's github pages are now online

sylikc committed:
Merge branch 'master' into v0.5.x-py3-refactor

# Conflicts:
#	CHANGELOG.md
#	exiftool/exiftool.py
#	setup.py

sylikc committed:
I messed up when merging PRs from GitHub... working with multiple remotes, sometimes forget to push all the right things to the right one.  Made this fix 2 weeks ago

Merge commit 'b4ab5f' into v0.5.x-py3-refactor

sylikc committed:
move the TestTagCopying to ExifToolHelper test script
fix TestTagCopying to use the TMP_DIR instead of creating (and leaving) an arbitrary test directory in the TEMP location of the system
fix bug in ExifToolHelper where passing in common_args=None will cause a crash

sylikc committed:
decided that execute_json() will return None if no output came back from execute()
fix helper to make sure get_tags() will detect None and raise an error

sylikc committed:
add in last_status.  This is the exit status from the previous command executed via exiftool
additional parameters and parsing was put in to read out the status code
fixed a bug that caused _parse_ver() to fail when return_tuple = True

sylikc committed:
update copyright on LICENSE file to reflect years as per standard

sylikc committed:
make the python.exe call explicit with .EXE extension in the batch files

sylikc committed:
add in a README which shows the requirements file for existing ones
rename pytest-cov.bat to pytest.bat as the requirements file will show the need for the -cov extension

sylikc committed:
exiftool.py fix all the:
* E261 at least two spaces before inline comment
* sys, codecs unused (that function was commented out)

sylikc committed:
fix the DeprecationWarning: Please use assertEqual instead.
add a new test for common_args

sylikc committed:
fix all: E261 at least two spaces before inline comment

sylikc committed:
changed ExifTool common_args=None means that an empty list is set.  A test verifies this behavior

sylikc committed:
fix most of the: E265 block comment should start with '# '

sylikc committed:
fix some more PEP8 stuff and other things caught in static analysis by flake8

sylikc committed:
adding restore of config file before doing the create/test/set that will come later

sylikc committed:
adding current flake8 configuration as per discussion on PEP8 coding style here: https://github.com/sylikc/pyexiftool/discussions/23#discussioncomment-795679

nyoungstudios committed:
doesn't use root logger

jangop committed:
Create lint-and-test.yml

sylikc committed:
made changes back in May, some extra comments and optimizing a few things
* the proc_args not to trigger another function call
* raise sends the errors up automatically

sylikc committed:
as per https://github.com/sylikc/pyexiftool/pull/24 , loguru is a possibility as a logger, so check the logger set differently.

Any drop-in replacement for logger will work with ExifTool

sylikc committed:
move the logger into the __init__ instead of global to exiftool.py

sylikc committed:
Merge branch 'nyoungstudios-master'

Merge the change as per https://github.com/sylikc/pyexiftool/pull/24 so that the proper getLogger() is used instead of the root logger

sylikc committed:
v0.4.11 changed documentation link to point to the right place in PyPI

also added the changelog entry for the last commit

sylikc committed:
v0.4.12 regarding pull request https://github.com/sylikc/pyexiftool/pull/26

didn't merge it directly as I would like to retain the outs, errs even if unused at this time

sylikc committed:
Merge branch 'master' into v0.5.x-py3-refactor

merge all the changes from v0.4.10, v0.4.11, v0.4.12

basically just documentation changes, but needed to get Git up to date

# Conflicts:
#	CHANGELOG.md
#	exiftool/exiftool.py
#	setup.py

sylikc committed:
ExifToolHelper - add auto_start feature.  Demonstrate overriding execute() to provide auto-start functionality

added comment about why exec_params is initialized that way in get_tags()

sylikc committed:
ExifToolHelper - fix bug that if you pass in common_args=[] it breaks because the check fails

sylikc committed:
ExifToolHelper - changed the way the constructor works

The way I had it implemented before was sort of dumb, and using **kwargs can pass on any and all stuff to ExifTool() constructor regardless of how it changes (logger was added from last time this was updated)

sylikc committed:
ExifToolHelper - support Path-like objects in get_tags()

Will have to do it in other functions as we move forward, but gotta start somewhere.

sylikc committed:
Exiftool main class code cleanup.
* remove Python 2.x "unicode_literals"
* clean up imports.  Remove unused and sort them by standard vs internal library ones
* remove the fscodec() commented out code
* rename internal function to start with underscore
* other misc comments

sylikc committed:
ExifToolHelper() - added terminate() overload

moved "not so well tested" functionality out to an ExifToolAlpha class, which will be an extension of ExifToolHelper

This code has not been tested yet.  The tests will be re-targeted to this multi layer class approach soon.

sylikc committed:
ExifTool main class
* remove "return_tuple" feature... I added it thinking it would be a useful feature, but in practice, it unnecessarily complicates the code paths
* consistently used f-strings.  removed all .format() uses
* clean up the way execute() sets its class variables.  Have all the local variables set in one spot, and class variables set much later, but together.  Cleaned up the names to be consistent.
* add in comments to sections of execute()

sylikc committed:
ExifToolHelper - fix technically incorrect comment... overload -> override

jangop committed:
Merge branch 'sylikc:master' into master

jangop committed:
Install exiftool

jangop committed:
Use apt-get instead of apt

jangop committed:
Install as root

jangop committed:
Assign IOError

I **guess** that this is what was meant.
Now, the variable `error` is redundant. 🤷

jangop committed:
Merge branch 'master' of github.com:jangop/pyexiftool

sylikc committed:
ExifTool - fix some bugs that caused actual use error... also, apparently "\" are not allowed anywhere in f-strings

sylikc committed:
IMPORTANT: API signature of get_tags() has been changed.. `files` and `tags` have been reversed

API signature of all future methods will change to have `files` first, `params` last
This should make it more consistent across the board as exiftool must work on "FILES" first and then options afterwards

Also fixed get_tags() to be generic enough that it won't require specifically type-checking against PurePath

jangop committed:
Represent all “files” as strings

I believe this list comprehension is what
the original author was looking for.

sylikc committed:
Merge branch 'jangop-v0.5.x-py3-refactor' into v0.5.x-py3-refactor

nice Pythonic 1-liner from @jangop https://github.com/sylikc/pyexiftool/pull/31

sylikc committed:
ExifTool() - added `encoding` constructor parameter (requires Python 3.6+), and R/W property

Modified all functionality to remove the use of bytes to communicate with the process, but instead use the encoding specified on Popen()

This appears to address the issue noted in here https://github.com/sylikc/pyexiftool/issues/29

sylikc committed:
Tests tests tests

All tests fixed to pass with the new encoding parameter.  I suspect because the encoding of the py files are UTF-8 the tests in this case need to use UTF-8 to get the same data back and forth

Deleted the "tests/tmp" dir and now it uses a tempfile that just gets wiped when it's done.

jangop committed:
Parse and compare versions of exiftool

sylikc committed:
* update documentation in ExifTool() class.
** removed some obsolete comment from the execute() method
* update comments in test_alpha to remind myself why I had to specify the temp_obj

sylikc committed:
ExifTool - add comment that ${status} only supported on exiftool v12.10+
ExifTool - change the cmd_text to use f-string which some users have documented to be way faster than concatenation

sylikc committed:
Examine PR https://github.com/sylikc/pyexiftool/pull/20

Merge branch 'basic-tests' of git://github.com/jangop/pyexiftool into jangop-basic-tests

sylikc committed:
Merge Pull request which adds some basic tests to the ExifToolHelper

https://github.com/sylikc/pyexiftool/pull/20

Merge branch 'jangop-basic-tests' into v0.5.x-py3-refactor

sylikc committed:
Merge branch 'v0.5.x-py3-refactor' of github.com:sylikc/pyexiftool into v0.5.x-py3-refactor

Forgot to push before merging the PR on the branch

sylikc committed:
Github workflows to do testing... check required and warn on others

Merge PR #30 by https://github.com/jangop

sylikc committed:
merged jangop changes but reverted some of the removed comments.  kept to his styling mostly

sylikc committed:
Merge PR https://github.com/sylikc/pyexiftool/pull/32 by https://github.com/jangop

Changes the way "version" is detected in tests by using packaging.version rather than a float comparison.

sylikc committed:
Pull the latest changes (GitHub workflows lint-and-test) from master into py3 refactor branch to get the same functionality.

sylikc committed:
Add versioning information __version__ to the package.
removed version= in the setup.py and moved it to the setup.cfg

sylikc committed:
invalid syntax fix

this bugfix was fixed in ecbd420acf4c18acf27d52d9c0b2728c829d7431 but didn't propogate when I did the last merge from master

sylikc committed:
* was in the process of adding a version check, but realized it may not be necessary as the current required 12.15 will fail on GETTING the version already, no need to check it again at this time.
* removed version_tuple() property as it was supposed to be for a future version check.  It's rolled into a private method which also got commented out as it's not needed at this time

This removes some brittle code (the version check is hacky), which can be added back later if a "version check" functionality is needed

sylikc committed:
update README.rst to prepare for initial Python3 release.  Removed outdated information and talked about the refactor in an easy-to-understand way.

sylikc committed:
test_helper.py
* removed the expected fail on initialize
* switched the parameters to match the new Helper() convention,
* added a TypeError test for get_tags()

sylikc committed:
Adding a GitHub workflow to make sure the minimum required version of exiftool is used rather than the default
prints out the version after install so we are sure what's going on

sylikc committed:
GitHub action - try again to get the PATH variable to the test steps

sylikc committed:
v0.4.13 - added a comment in CHANGELOG.md for the changes.  This was a bugfix release and should be the last Python2 release.  This will be published on PyPI

sylikc committed:
v0.4.13 - forgot to bump the version on setup.py

sylikc committed:
merge the v0.4.13 changelog in from master one last time

sylikc committed:
Update year on LICENSE

sylikc committed:
Add a "Brief History" section to README for some insight to how all of this came to be.  Turned it into UTF-8

sylikc committed:
ExifTool - changed the win_shell default (I actually am not sure this feature works right now)
updated some of the comments and sample code to match what currently would be

sylikc committed:
updated COMPATABILITY.txt with some of the changed features.  It's not pretty at the moment, but it should have the core changes which v0.5.0 provides

sylikc committed:
Update README.rst
trying to format better,
added in the special thanks,
and reordered some sections/split into subsections

sylikc committed:
more refinements to README.rst

sylikc committed:
ExifToolHelper - changed files invalid files parameter to a ValueError
  while testing, realized that an invalid params would cause weird problems, so added error checking to params in Helper

test_helper.py - added test and fixed failing tests

sylikc committed:
test_exiftool.py - added a few tests to increase code coverage

sylikc committed:
update .gitignore as per https://github.com/sylikc/pyexiftool/pull/39

the output of tests changed and ignore some IDE stuff

sylikc committed:
ExifTool
* remove -w no output detection.  If a user uses it, they shouldn't be calling execute_json()
* remove some dead code and change code flow of execute_json() as per https://github.com/sylikc/pyexiftool/pull/39

sylikc committed:
.gitignore changed again... not sure why /tmp/ was added in the PR

sylikc committed:
ExifTool / ExifToolHelper
* Add in exceptions classes and made all tests pass.  Added a -w test.

ExifToolHelper override run() to remove warning just like terminate()

sylikc committed:
forgot to import xceptions into helper

sylikc committed:
ExifTool
* added tests for config_file and setting it to a valid one
* config_file can be "" as per documentation, so changed code to make "" valid
* added additional typing indicators

sylikc committed:
TestExifToolHelper
* fixed bug which created random temp directories int he wrong place

sylikc committed:
* ExifTool - add some typing and documentation updates (better docstrings)
* COMPATIBILITY.txt - beautify it to make it clearer... might still have more updates after deployment
* CHANGELOG.md - add in the line to deploy which I've been waiting for for a year...

sylikc committed:
Merge the Python3 Refactor code.  Almost a complete re-write of non-core functionality.

This branch has been active on GitHub a year in the making... so there's probably too many changes to review based on commits (128 total commits on branch).

Read COMPATIBILITY.txt and browse the code to understand the full extent of the changes.  Documentation updates are coming soon.

Merge branch 'v0.5.x-py3-refactor'

sylikc committed:
Fix README.rst to make it compatible with twine
minor changes to documentation / dates.

sylikc committed:
Documentation update.  Every single docstring was touched (except ExifToolAlpha).

Sphinx documentation has been updated and will be updated on github.io with a version matching this commit

sylikc committed:
add the inheritance diagram to the Class pages, and tweaked the output to be a bit cleaner

sylikc committed:
Tweaked the README a bit and now there's one less duplicate blurb in the documentation

sylikc committed:
added a new exception ExifToolExecuteError() which will be returned on a non-zero exit status

ExifToolHelper() added a check parameter to execute/execute_json/get_tags.  (not yet decided, but currently defaulting the lower methods to False, and the higher level methods to True

sylikc committed:
ExifTool - no functional changes
* added some more type hints
* Updated documentation for execute() and execute_json()

sylikc committed:
ExifToolAlpha
* moved set_tags_batch() and set_tags() out to Helper, with different method signature

ExifToolHelper
* add more type hints and better documentation
* partial undo the check= parameter from last commit... making it a check_execute flag at the constructor level
* added a property to get/set check_execute property
* removed the execute_json() override.  Tested the code paths and it's an unnecessary override (without it, the same behavior happens)
* added set_tags method which combines the methods from ExifToolAlpha.  Similar signature to get_tags() - I figured this should be the most requested functionality missing in v0.5.0

sylikc committed:
updated tests for ExifToolHelper and ExifToolAlpha for coverage of the new methods and properties

sylikc committed:
README.rst - adding some badges to make it look cooler
setup.py - ignore the badges when using README.rst as PyPI description

sylikc committed:
add a Windows batch file to the scripts directory so I can get consistent docs generation results

sylikc committed:
ExifToolExecuteError - add a better docstring to indicate that there are attributes which can be read from this error object

sylikc committed:
update the scripts README.txt file so it better reflects the current usage of scripts, now that I'm also doing the docs building from there

sylikc committed:
ExifToolHelper
* remove extraneous print statement (was used to test the execute() codepath)
* add a warning regarding disabling check_execute

sylikc committed:
* moved some tests from test_alpha to test_helper
* added a few new tests to ensure that the ExifToolExecuteError is working as it should
* added test to set keywords with set_tags() with a list
* removed the use of str(Path) in the tests, as the get_tags/set_tags supports Path directly

sylikc committed:
docs changes
* remove private-members and imported-members from autoapi (less duplicate and not-useful info)
* explicitly add autoapi docs to the end, so the toc displays the submembers properly
* moved the PyExifTool docstring from exiftool.exiftool to exiftool.__init__ file.  Makes more sense...
* added a docstring note to exiftool/helper/experimental that the class is available at the base namespace

sylikc committed:
__init__.py removed the DEFAULT_EXECUTABLE from the main namespace.  Users should not have to ever interface with this this variable.

updated CHANGELOG.md with the changes ready for the v0.5.2 release

sylikc committed:
setup.py bugfix - forgot to change the variable to the function

sylikc committed:
ExifToolHelper - fixed a typo in the docstring and changed the wording of the warning for wildcards

sylikc committed:
no functional changes

* ExifToolHelper - added 3 static helper methods to deal with the parsing of arguments which is common to get_tags() and set_tags()
(the _check_tag_list is not functional yet)
* ExifToolAlpha - change set_keywords_batch to demonstrate how it can use one of the helper functions from ExifToolHelper to take advantage of the common functionality

sylikc committed:
no functional changes
* Update LICENSE and standardize the license description on the top of the main files
* minor docstring updates

sylikc committed:
* Update tests to use a common method to create the temp_dir (centralizes duplicate code)
* Don't create the temp_dir on all methods, only on demand... maybe group methods later
* Don't delete files created in tests, for analysis later if temp_dir persists

sylikc committed:
README.rst add table of contents ... the text has gotten long and difficult to navigate on GH homepage

sylikc committed:
REAMDE.rst remove the backlinks, and tweak slightly.  This should make navigation on the GitHub rendered page much easier

sylikc committed:
README.rst Looks like I have to explicitly disable backlinks

sylikc committed:
* setup.py - add some classifiers + change project status from Alpha to Beta... I'd say the past year of updates makes this more beta quality (except for ExifToolAlpha of course)
* README.rst - going crazy with badges!

sylikc committed:
ExifToolAlpha - made the get_tag way more robust with checks... left get_tag_batch() alone though

sylikc committed:
* exceptions - added ExifToolTagNameError
* ExifToolHelper - added the check_tag_names property and associated _check_tag_list() but it's still a bit wonky right now, exiftool allows too many variations of patterns...

may not leave default to check... might make it an option.

wrote associated tests

sylikc committed:
ExifTool
* execute() debug log is clearer as to where it came from, what is input and output
* added in logging of the input params
* renamed the errors as per change in exceptions
* added some documentation as to why errors are not checked in the base ExifTool class
exceptions
* rename a few exceptions and created a base class for anything execute() related so it can return a common object
* refactor with slightly more comments
* rename of some of the exceptions to clearly indicate the source is ExifTool

sylikc committed:
All tests pass again

fix ExifToolHelper imports for renamed errors
fix test_helper's imports so all tests pass again
fix import in ExifToolAlpha

sylikc committed:
refactor the tests directory to break down by files, so imports are cleaner, and it's easier to find which tests i'm looking at

sylikc committed:
tests - add a few more statements to get coverage, and move a set_tags test to the right file

sylikc committed:
trying to update the GitHub Action to use a cache for the exiftool download

sylikc committed:
forgot to end the if with fi

sylikc committed:
the job at least ran successfully... so turn on caching for pip as well

sylikc committed:
I <3 tabs... github doesn't though

sylikc committed:
can't enable python cache without using a requirements file

sylikc committed:
Merge in the github actions which now cache the downloaded ExifTool source (from perl.org)

This should speed up the build and save on resources as GitHub doesn't have to re-download the file every time!

Merge branch 'gh_actions'

sylikc committed:
no functional changes.  some docstring updates.

sylikc committed:
no functional changes.  Added a "Docs" badge as per suggestion https://github.com/sylikc/pyexiftool/discussions/40#discussioncomment-2389111 ... also simplified the wording for the setup-from-source instructions

sylikc committed:
Fixed ujson compatibility.  ujson isn't a "drop-in" replacement after all

sylikc committed:
no functional changes.  Update documentation slightly, and preparing for 0.5.3 release

sylikc committed:
no functional changes.  Updated some docstrings.  Added Examples and FAQ page.  Updated docs with a link to GitHub

sylikc committed:
no functional changes.  A few minor adjustments to examples documentation

sylikc committed:
no functional changes.  Added docs that execute() can throw UnicodeDecodeError.  Updated docs on constructor

sylikc committed:
tweaked the conversion of params to text, the logging doesn't need to use .split() on a .join() ... just use the variable

sylikc committed:
remove the `encoding` parameter to the call of Popen()

add in the raw_bytes parameter to execute() ... tests currently pass if raw_bytes is false.

sylikc committed:
rename the "images" folder to "files" folder as we're going to create a config file to test the new raw_bytes feature

sylikc committed:
adjust the static typing for the new possible return types from the usage of execute().  update the docstring as well

sylikc committed:
ExifTool fix bug that was introduced in status code checking when raw_bytes=True

sylikc committed:
added a ExifTool bytes read/write test

sylikc committed:
ExifTool - optimized memory of params... instead of creating the copy of the params list twice, make only one copy (like it was originally)

sylikc committed:
ExifToolHelper - add in a helper at the cost of making another copy of the list of parameters... allows passing in all arbitrary objects that just get str()

sylikc committed:
ExifToolHelper - fixed bug that the params didn't all get str()'d properly... now all tests updated to test this behavior

sylikc committed:
Docstring updates, and examples and FAQ updates

sylikc committed:
Write a coverage test (and test that the last_stdout and last_stderr) are available after terminate

sylikc committed:
ExifToolHelper
* update _is_iterable() to have an ignore_str_bytes parameter
* execute() uses list comprehension instead of the if-else.  Renamed the params list to make it clearer
* get_tags() and set_tags() files/params are allowed to be arbitrary objects.  let execute() handle the casting

Add tests to get full coverage.  Fix the params test before

sylikc committed:
fix the wrong path in the FAQ due to the restructuring

sylikc committed:
no functional changes.  bump version for 0.5.4 release.  update changelog

sylikc committed:
fix documentation about "utf-8" reported in issue #58

prutschman committed:
Speed operations on large responses

Keep a list of bytes instead of repeatedly appending (copying) to the bytes
object.

sylikc committed:
modified a bit of the styling using double-quotes instead of single-quotes (when possible, like not with f-strings)... then fixed a bit of previous code with the same styling difference

sylikc committed:
Change workflow to explicitly use ubuntu-20.04 instead of ubuntu-latest to (lazy) fix the 3.6 arch x64 problem.

I think there's other better fixes, but the thread is super long https://github.com/actions/setup-python/issues/544

sylikc committed:
add in python 3.10 and 3.11 into the test, since it should be supported

sylikc committed:
Speed operations on large responses.

This PR addresses the append() operations and uses lists and reverse() to speed up getting exiftool output

Merge PR https://github.com/sylikc/pyexiftool/pull/61 by https://github.com/prutschman

sylikc committed:
Attempt to fix the github action for 3.10 becoming 3.1 by using quotes

https://github.com/actions/setup-python/issues/160

sylikc committed:
Wrong indent level for the example in README.rst

https://github.com/sylikc/pyexiftool/pull/57 thanks Sylvain Poulain https://github.com/kikislater for catching that!

sylikc committed:
update typing for execute_json() and some documentation changes related to 0.5.4 update

sylikc committed:
Update version to 0.5.5 and update CHANGELOG to reflect changes.

No functional changes, only a speed optimization of existing code.

sylikc committed:
ExifTool - remove automatic ujson support, but support setting a custom json.loads() - documentation not complete yet for method, and naming is not final
ExifToolHelper - updated some documentation.

sylikc committed:
some light cleanup on ExifToolAlpha

sylikc committed:
workflow add in [json] to install "supported" processors

sylikc committed:
add test for float in tags testing, which uses the set_json_loads() method the way it's intended, to workaround exiftool JSON float handling glitch

sylikc committed:
no functional changes:
* add documentation for the set_json_loads() metdho
* Fix some flake8 styling and unused variables linting

sylikc committed:
Fix a few typing errors and made sure unittest runs properly

sylikc committed:
add additional test cases to the float test

sylikc committed:
update documentation and write two FAQ articles as to how to use the new method

sylikc committed:
no functional changes
* update all copyright years to 2023
* update CHANGELOG.md for 0.5.6 release
* prepare for release

sylikc committed:
Merge the changes to support ExifTool.set_json_loads() for final 0.5.6 release

sylikc committed:
in CPython 3.12, instead of freezing in ExifTool.terminate(), it crashes.  Add a try/except so it works properly in CPython 3.12.0

sylikc committed:
Documentation updates only.  No functional changes.

FAQ
- fix a typo in set_json_loads() FAQ
- add a section in FAQ to use shlex to help debug argument splitting.  Lots of issues are opened just on how to split arguments.
- refer to the FAQ from the quickstart/examples page

small tweak to the changelog

sylikc committed:
Documentation and formatting updates.  No functional changes.

helper.py TUPLE_STR_BYTES, while functionally the same, should be str, bytes

constants.py, put DEFAULT_EXECUTABLE into a one-liner so that it's properly documented by AutoAPI

sylikc committed:
add Python 3.12 to GitHub actions
removed accidentally added file from a few commits ago (no need to rewrite history as it's just a random text file with links to resources)
